pkg/pillar/cmd/zedagent/handlemetrics.go

```diff
@@ -528,6 +531,50 @@ func publishMetrics(ctx *zedagentContext, iteration int) {
 	}
 	log.Tracef("DirPaths in persist, elapse sec %v", time.Since(startPubTime).Seconds())
 
+	// ZFS metrics
+	if vault.ReadPersistType() == types.PersistZFS {
+		zpoolList, err := libzfs.PoolOpenAll()
+		if err != nil {
+			log.Errorf("get zpool list failed %v", err)
+		} else {
+			for _, zpool := range zpoolList {
+				rStorageMetric := new(zmet.StorageMetric)
+				rStorageMetric.PoolName = zpool.Properties[libzfs.PoolPropName].Value
+				rStorageMetric.StorageState = zfs.GetZfsDeviceStatusFromStr(zpool.Properties[libzfs.PoolPropHealth].Value)
+				rStorageMetric.CountVolume, err = zfs.GetZfsCountVolume(zpool.Properties[libzfs.PoolPropName].Value)
+				if err != nil {
+					log.Errorf("get count volume failed %v", err)
+				}
+
+				vdevs, err := zpool.VDevTree()
+				if err != nil {
+					log.Errorf("get info about disks in zpool failed %v", err)
+				} else {
+					for _, vdev := range vdevs.Devices {
+						// If this is a RAID or mirror, look at the disks it consists of
+						if vdev.Type == libzfs.VDevTypeMirror || vdev.Type == libzfs.VDevTypeRaidz {
+							for _, disk := range vdev.Devices {
+								rDiskStatus, err := zfs.GetZfsDiskAndStatus(disk)
```

Comment:
Don't you want to log an error?

Reply:
In this case, no, because in the function that is called here, a check is made that the object is of the "Disk" type, and if it is not a disk, then in this case I do not add it.

Reply:
But why would dev.Devices contain things which are not disks?

Reply:
Because VDevTree struct reflects the type of any [device in the pool](https://github.com/bicomsystems/go-libzfs/blob/v0.4.0/common.go#L37), and some types of which may have their own devices.
In this case, the logic is quite simple, if we do not have a raid, then we must collect information about disks in the current type, if there is still a raid, then we will collect information about disks from it.